main.go

```diff
@@ -68,93 +71,130 @@ func main() {
 	// improving concurrency performance across multiple goroutines.
 	runtime.GOMAXPROCS(numCores)
 
-	shardManager := shard.NewShardManager(int8(numCores), watchChan, logr)
-
-	// Initialize the AsyncServer
-	asyncServer := server.NewAsyncServer(shardManager, watchChan, logr)
-	httpServer := server.NewHTTPServer(shardManager, watchChan, logr)
-
-	// Initialize the HTTP server
-
-	// Find a port and bind it
-	if err := asyncServer.FindPortAndBind(); err != nil {
-		cancel()
-		logr.Error("Error finding and binding port",
-			slog.Any("error", err),
-		)
-		os.Exit(1)
-	}
+	// Initialize the ShardManager
+	shardManager := shard.NewShardManager(uint8(numCores), watchChan, serverErrCh, logr)
 
 	wg := sync.WaitGroup{}
-	// Goroutine to handle shutdown signals
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		<-sigs
-		asyncServer.InitiateShutdown()
-		cancel()
-	}()
 
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
 		shardManager.Run(ctx)
 	}()
 
-	serverErrCh := make(chan error, 2)
 	var serverWg sync.WaitGroup
-	serverWg.Add(1)
-	go func() {
-		defer serverWg.Done()
-		// Run the server
-		err := asyncServer.Run(ctx)
 
-		// Handling different server errors
-		if err != nil {
-			if errors.Is(err, context.Canceled) {
-				logr.Debug("Server was canceled")
-			} else if errors.Is(err, server.ErrAborted) {
-				logr.Debug("Server received abort command")
+	// Initialize the AsyncServer server
+	// Find a port and bind it
+	if !config.EnableMultiThreading {
+		asyncServer := server.NewAsyncServer(shardManager, watchChan, logr)
+		if err := asyncServer.FindPortAndBind(); err != nil {
+			cancel()
+			logr.Error("Error finding and binding port", slog.Any("error", err))
+			os.Exit(1)
+		}
+
+		serverWg.Add(1)
+		go func() {
+			defer serverWg.Done()
+			// Run the server
+			err := asyncServer.Run(ctx)
+
+			// Handling different server errors
+			if err != nil {
+				if errors.Is(err, context.Canceled) {
+					logr.Debug("Server was canceled")
```

Comment:
Since it is an error, shouldn't we use logr.Error() ? And also, rather that logging **Server was cancelled** as the error message, can't we use **Server context was cancelled** ? Also, is there a way to know by which means context is cancelled ? Is it due to timeout, shutdown signal, manual cancellation etc ? Using the reason of context cancellation in the error message will be more useful.

Reply:
Again, this code existed earlier. I just re-organized the code for multi-threading use-case. 
Let us address thus as part of a separate refactoring PR.